krun/results.py

```diff
@@ -0,0 +1,106 @@
+        self.eta_estimates = dict()
+
+        # error_flag is flipped when a (non-fatal) error or warning occurs.
+        # When Krun finishes and this flag is true, a message is printed,
+        # thus prompting the user to investigate.
+        self.error_flag = False
+
+        # Fill in attributes from the config, platform and prior results.
+        if self.config is not None:
+            self.filename = self.config.results_filename()
+            self.init_from_config()
+        if platform is not None:
+            self.starting_temperatures = platform.starting_temperatures
+            self._audit = Audit(platform.audit)
+        else:
+            self.starting_temperatures = list()
+            self.audit = dict()
+
+        # Import data from a Results object serialised on disk.
+        if results_file is not None:
+            self.read_from_file(results_file)
+
+    @property
+    def audit(self):
+        return self._audit
+
+    @audit.setter
+    def audit(self, audit_dict):
+        self._audit = Audit(audit_dict)
+
+    def init_from_config(self):
+        """Scaffold dictionaries based on a given configuration.
+        """
+        # Initialise dictionaries based on config information.
+        for vm_name, vm_info in self.config.VMS.items():
+            for bmark, _ in self.config.BENCHMARKS.items():
+                for variant in vm_info["variants"]:
+                    key = ":".join((bmark, vm_name, variant))
+                    self.data[key] = []
+                    self.eta_estimates[key] = []
+
+    def read_from_file(self, results_file):
+        """Initialise object from serialised file on disk.
+        """
+        with bz2.BZ2File(results_file, "rb") as f:
+            results = json.loads(f.read())
+            self.__dict__.update(results)
+            # Ensure that self.audir and self.config have correct types.
+            self.config = Config(None)
+            self.config.read_from_string(results["config"])
+            self.audit = results["audit"]
+
+    def write_to_file(self):
+        """Serialise object on disk.
+        """
+        to_write = {
+            "config": self.config.text,
+            "data": self.data,
+            "audit": self.audit.audit,
+            "reboots": self.reboots,
+            "starting_temperatures": self.starting_temperatures,
+            "eta_estimates": self.eta_estimates,
+            "error_flag": self.error_flag,
+        }
+        with bz2.BZ2File(self.filename, "w") as f:
+            f.write(json.dumps(to_write,
+                               indent=1, sort_keys=True, encoding='utf-8'))
+
+    def jobs_completed(self, key):
+        """Return number of executions for which we have data for a given
+        benchmark / vm / variant triplet.
+        """
+        return len(self.data[key])
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return False
+        return (self.config == other.config and \
```

Comment:
Not sure if this interests you: `krun/results.py|100 col 49| E502 the backslash is redundant between brackets`